### PR TITLE
Support recreating model entities

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary

This PR adds support for recreating model entities. This feature is implemented to for the model editor work. Once a user makes changes to a model, i.e. by updating its or components or it's child entities' components, the idea is to just delete the old model and create a new one. This avoids having to go to different systems, detect what changed, and apply updates.

It introduces a new `components::Recreate` component. When attached to an model entity, the `SimulationRunner` will mark the existing model (recursively) as removed and recreate a new model by cloning from the original one. The cloned / recreated model will have the same name but different Entity Ids.

## Test it

Added a `INTEGRATION_recreate_entities` test

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by nkoenig** [comment]
Does the integration test pass for you? I'm getting failures:
```
[Err] [EntityComponentManager.cc:431] Requested to clone entity [5] with a name of [box_link], but another entity already has this name.
[Err] [EntityComponentManager.cc:552] Cloning child entity [5] failed.
[Err] [EntityComponentManager.cc:431] Requested to clone entity [9] with a name of [cylinder_link], but another entity already has this name.
[Err] [EntityComponentManager.cc:552] Cloning child entity [9] failed.
[Err] [EntityComponentManager.cc:431] Requested to clone entity [13] with a name of [sphere_link], but another entity already has this name.
[Err] [EntityComponentManager.cc:552] Cloning child entity [13] failed.
[Err] [EntityComponentManager.cc:431] Requested to clone entity [17] with a name of [capsule_link], but another entity already has this name.
[Err] [EntityComponentManager.cc:552] Cloning child entity [17] failed.
[Err] [EntityComponentManager.cc:431] Requested to clone entity [21] with a name of [ellipsoid_link], but another entity already has this name.
[Err] [EntityComponentManager.cc:552] Cloning child entity [21] failed.
/home/nkoenig/work/ignition/fortress/ign-gazebo/test/integration/recreate_entities.cc:134: Failure
Expected equality of these values:
  24u
    Which is: 24
  ecm->EntityCount()
    Which is: 4
```

**Comment by iche033** [comment]
> Does the integration test pass for you? I'm getting failures:

fixed in 39a1d34

**Comment by chapulina** [comment]
> The cloned / recreated model will have the same name but different Entity Ids.

Is the idea that systems will detect that the model was removed with `EachRemoved` and then create a new one with `EachNew`? Does removal and creation happen in the same iteration? This may cause issues if a system calls `EachNew` before `EachRemoved`.

**Comment by iche033** [comment]
the order of entity removal and creation is done here:
https://github.com/ignitionrobotics/ign-gazebo/blob/68faa5acbf83b27d9bdebe23cba3e78051741817/src/SimulationRunner.cc#L836-L847

While the remove and create requests are done in the same sim iteration, the systems will detect entities removed in the current iteration then re-recreate the entity in the next iteration. This is done to avoid creating duplicate entities when `EachNew / Each` is called before `EachRemoved`

File: src/SimulationRunner.cc [hunk 1]
```diff
@@ -1252,6 +1264,52 @@ void SimulationRunner::ProcessWorldControl()
...
+    Entity clonedEntity = this->entityCompMgr.Clone(ent,
+       parentComp->Data(), nameComp->Data(), false);
+    clonedEntities.insert(clonedEntity);
+
+  }
+  // remove the Recreate component so they do not get recreated again in the
+  // next iteration
+  for (auto &ent : clonedEntities)
+  {
+    this->entityCompMgr.RemoveComponent<components::Recreate>(ent);
```
Comment (chapulina) [inline src/SimulationRunner.cc:1307]: Can't this be done in the loop above?

**Comment by nkoenig** [comment]
Retarted to the `model_editor` branch.

**Comment by nkoenig** [comment]
@iche033, I could use some help debugging this functionality. See #1187.

File: src/SimulationRunner.cc (hunk 1 above)
Comment (nkoenig) [reply src/SimulationRunner.cc]: Indeed. Moved in 14d78f3

File: src/SimulationRunner.cc [hunk 2]
```diff
@@ -1252,6 +1264,52 @@ void SimulationRunner::ProcessWorldControl()
...
+/////////////////////////////////////////////////
+void SimulationRunner::ProcessRecreateEntitiesRemove()
+{
+  IGN_PROFILE("SimulationRunner::ProcessRecreateEntitiesRemove");
+
+  // store the original entities to recreate and put in request to remove them
+  this->entityCompMgr.Each<components::Model,
+                           components::Recreate>(
+      [&](const Entity &_entity,
+          const components::Model *,
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/SimulationRunner.cc:1300]: Added documentation in 14d78f3

File: include/ignition/gazebo/components/Recreate.hh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace components
+{
+  /// \brief A component that identifies an entity needs to be recreated
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply include/ignition/gazebo/components/Recreate.hh]: Added documentation in 14d78f3

File: src/SimulationRunner.cc [hunk 4]
```diff
@@ -1252,6 +1264,52 @@ void SimulationRunner::ProcessWorldControl()
...
+{
+  IGN_PROFILE("SimulationRunner::ProcessRecreateEntitiesCreate");
+
+  // clone the original entities
+  std::set<Entity> clonedEntities;
+  for (auto & ent : this->entitiesToRecreate)
+  {
+    auto nameComp = this->entityCompMgr.Component<components::Name>(ent);
+    auto parentComp =
+        this->entityCompMgr.Component<components::ParentEntity>(ent);
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/SimulationRunner.cc:1319]: Added checks in 14d78f3

**Review by nkoenig — APPROVED** [review]

**Comment by iche033** [comment]
nice, I was able to recreate a model with sensors in the links. 

When I tried to recreate a rendering sensor, e.g. rgbd camera, I get an error msg that the camera already exists. This should be fixed by https://github.com/ignitionrobotics/ign-gazebo/pull/1170/commits/8a6552e589935e5830bec5eeafdeff8e65359c46 and https://github.com/ignitionrobotics/ign-sensors/pull/169

**Comment by nkoenig** [comment]
I'll merge this once CI looks okay.

**Comment by mjcarroll** [comment]
Out of curiosity, did you try this on any models with joints?  If not, I can.

**Comment by nkoenig** [comment]
Oh I did not try this on a model with joints.

**Comment by mjcarroll** [comment]
Just did a simple test world with two double pendulums here: https://github.com/ignitionrobotics/ign-gazebo/tree/recreate_entities_joints

I'm getting errors like

```
[Err] [EntityComponentManager.cc:1982] Error: attempted to clone links, but link [14] was never cloned.
[Err] [EntityComponentManager.cc:388] Error updating the cloned parent link name for cloned joint [92]
[Err] [EntityComponentManager.cc:1982] Error: attempted to clone links, but link [9] was never cloned.
[Err] [EntityComponentManager.cc:388] Error updating the cloned parent link name for cloned joint [91]
```

**Comment by nkoenig** [comment]
> Just did a simple test world with two double pendulums here: https://github.com/ignitionrobotics/ign-gazebo/tree/recreate_entities_joints
> 
> I'm getting errors like
> 
> ```
> [Err] [EntityComponentManager.cc:1982] Error: attempted to clone links, but link [14] was never cloned.
> [Err] [EntityComponentManager.cc:388] Error updating the cloned parent link name for cloned joint [92]
> [Err] [EntityComponentManager.cc:1982] Error: attempted to clone links, but link [9] was never cloned.
> [Err] [EntityComponentManager.cc:388] Error updating the cloned parent link name for cloned joint [91]
> ```

Is this something you are looking into, or should I take it on?

**Comment by mjcarroll** [comment]
I'm looking into it.  It reliably happens when I have two double pendulums that are cloned, but not if there is a single model in the world.

**Comment by mjcarroll** [comment]
@nkoenig see #1206

**Pull request merged**